pkg/app/ops/insightcollector/collector.go

```diff
@@ -62,119 +65,123 @@ var (
 	pageSize = 50
 )
 
-func (i *InsightCollector) CollectProjectsInsight(ctx context.Context) error {
-	now := time.Now().UTC()
-	maxCreatedAt := now.Unix()
-
-	for {
-		projects, err := i.projectStore.ListProjects(ctx, datastore.ListOptions{
-			PageSize: pageSize,
-			Filters: []datastore.ListFilter{
-				{
-					Field:    "CreatedAt",
-					Operator: "<",
-					Value:    maxCreatedAt,
-				},
-			},
-			Orders: []datastore.Order{
-				{
-					Field:     "CreatedAt",
-					Direction: datastore.Desc,
-				},
-			},
-		})
-		if err != nil {
-			return err
-		}
-		if len(projects) == 0 {
-			// updated all project's insights completely
-			break
+func (i *InsightCollector) AggregateWithCreatedAt(ctx context.Context) error {
+	currentTime := time.Now()
+	targetDate := time.Date(currentTime.Year(), currentTime.Month(), currentTime.Day(), 0, 0, 0, 0, time.UTC)
+	m, err := i.insightstore.LoadMilestone(ctx)
+	if err != nil {
+		if err == filestore.ErrNotFound {
+			m = &insight.Milestone{}
 		}
+		return err
+	}
 
-		for _, p := range projects {
-			for _, k := range aggregateKinds {
-				if err := i.updateApplicationChunks(ctx, p.Id, "", k, now); err != nil {
-					i.logger.Error("failed to update application chunks", zap.Error(err))
-				}
+	dc, err := i.getDeploymentsWithCreatedAt(ctx, m.DeploymentCreatedAtMilestone, targetDate.Unix())
+	if err != nil {
+		return err
+	}
+	appMap, projectMap := i.groupingDeployments(dc)
+
+	var returnErr error
+	for appID, ds := range appMap {
+		for _, k := range metricsAggregateWithCreatedAt {
+			if err := i.updateApplicationChunks(ctx, ds[0].ProjectId, appID, ds, k, targetDate); err != nil {
+				i.logger.Error("failed to update application chunks", zap.Error(err))
+				returnErr = err
 			}
 		}
-		maxCreatedAt = projects[len(projects)-1].CreatedAt
 	}
-	return nil
-}
+	for proID, ds := range projectMap {
+		for _, k := range metricsAggregateWithCreatedAt {
+			if err := i.updateApplicationChunks(ctx, proID, ds[0].ApplicationId, ds, k, targetDate); err != nil {
+				i.logger.Error("failed to update application chunks", zap.Error(err))
+				returnErr = err
+			}
+		}
+	}
+	if returnErr == nil {
+		m.DeploymentCreatedAtMilestone = targetDate.Unix()
+	}
+	if err := i.insightstore.PutMilestone(ctx, m); err != nil {
```

Comment:
Do we need to update the milestone when `returnErr` was not nil?

Reply:
Exactly :+1: